### PR TITLE
RES-128 Hierarchical clustering implementation

### PR DESCRIPTION
**Pull request by SaganBolliger**
https://jira.numenta.com/browse/RES-128

WIP

@BoltzmannBrain @subutai Comments welcome.

### CODE REVIEW

File: htmresearch/algorithms/hierarchical_clustering.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+
+
+class HierarchicalClustering(object):
+  """
+  Implements hierarchical agglomerative clustering on the output of a classification network.
+  """
+
+
+  def __init__(self, classificationNetwork, linkageMethod='single'):
```
Comment (subutai) [inline htmresearch/algorithms/hierarchical_clustering.py]: Perhaps change the parameter to be an instance of type knn. The experiment runner can extract the KNN from the classificationNetwork and pass it in.  That would make this class a bit more general.

File: htmresearch/algorithms/hierarchical_clustering.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    for i in xrange(knn._numPatterns):
+      nzIndices = knn.getPattern(i, sparseBinaryForm=True)
+      sparseRow = scipy.sparse.csr_matrix(
+        (numpy.ones(len(nzIndices), dtype=bool),
+        (numpy.zeros(len(nzIndices)), nzIndices)),
+        shape=(1,dim))
+      sparseRowList.append(sparseRow)
+
+    sparseDataMatrix = scipy.sparse.vstack(sparseRowList)
+    overlaps = HierarchicalClustering._computeOverlaps(sparseDataMatrix)
```
Comment (subutai) [inline htmresearch/algorithms/hierarchical_clustering.py]: This and the subsequent call to cluster could take a while. Typically constructors should return fast. Perhaps put the heavy weight code in a separate method that is explicitly called?

File: htmresearch/algorithms/hierarchical_clustering.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    print overlaps.nbytes / 1024**2 # MB
```
Comment (subutai) [inline htmresearch/algorithms/hierarchical_clustering.py]: Remove print statements.

File: htmresearch/algorithms/hierarchical_clustering.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    @param selfOverlaps (boolean) If true, include diagonal (density) values
+        from pairwise similarity matrix. Then returned vector has n(n+1)/2
+        elements. Optional, defaults to False.
+    
+    @param dtype (string) Data type of returned array in numpy dtype format.
+        Optional, defaults to 'int16'.
+    
+    @returns (numpy.ndarray) A vector of pairwise overlaps as described above
+        with dtype uint16.
```
Comment (subutai) [inline htmresearch/algorithms/hierarchical_clustering.py]: Not always uint16 - tied to the dtype parameter.

File: htmresearch/algorithms/hierarchical_clustering.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    linkage = scipy.cluster.hierarchy.linkage(-overlaps, method=linkageMethod)
+    
+    linkage[:,2] -= numpy.min(linkage[:,2])
+    print numpy.max(linkage[:,2])
+    self._linkage = linkage
+
+
+  @staticmethod
+  def _computeOverlaps(data, selfOverlaps=False, dtype='int16'):
```
Comment (subutai) [inline htmresearch/algorithms/hierarchical_clustering.py]: How do you know this method works correctly? Do you have any tests?

File: htmresearch/algorithms/hierarchical_clustering.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    return fig
+    
+
+
+if __name__ == "__main__":
+  knn = KNNClassifier()
+  for i in xrange(3500):
+    knn.learn(numpy.random.rand(50) < 0.2, 2)
+
+  class ClassifierNetwork(object):
```
Comment (subutai) [inline htmresearch/algorithms/hierarchical_clustering.py]: Unnecessary with suggested change to constructor.

File: htmresearch/algorithms/hierarchical_clustering.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+  def getDendrogram(self):
+    linkage = self.getLinkageMatrix()
+    fig = plt.figure()
+    ax = plt.axes()
+    scipy.cluster.hierarchy.dendrogram(linkage, p=30, truncate_mode="lastp", ax=ax)
+    return fig
+    
+
+
+if __name__ == "__main__":
```
Comment (subutai) [inline htmresearch/algorithms/hierarchical_clustering.py]: I would recommend moving this code block to projects/nlp/clustering as simple_clustering_example.py

**Comment by subutai** [comment]
@SaganBolliger Thanks - left some relatively minor suggestions. Overall looks good to me.

File: htmresearch/algorithms/hierarchical_clustering.py (hunk 7 above)
Comment (BoltzmannBrain) [inline htmresearch/algorithms/hierarchical_clustering.py]: It looks like this code was used for your debugging. Please remove it.

**Comment by BoltzmannBrain** [comment]
@SaganBolliger looks good overall, but please include some tests, which could also serve as good example cases. And just to be nitpicky: https://github.com/numenta/nupic/wiki/Python-Style-Guide#coding-style

**Comment by SaganBolliger** [comment]
@BoltzmannBrain Thanks Alex. Will be removing the debugging code. Were there particular issues with style that you noticed? I was trying to be compliant, except in the case of URLs in the doc strings that are over 80 chars (not sure if there's a better way to handle that).

File: htmresearch/algorithms/hierarchical_clustering.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    @param dtype (string) Data type of returned array in numpy dtype format.
+        Optional, defaults to 'int16'.
+    
+    @returns (numpy.ndarray) A vector of pairwise overlaps as described above
+        with dtype uint16.
+    """
+    nVectors = data.shape[0]
+    nDims = data.shape[1]
+    nPairs = (nVectors+1)*nVectors/2 if selfOverlaps else (
+        nVectors*(nVectors-1)/2)
```
Comment (BoltzmannBrain) [inline htmresearch/algorithms/hierarchical_clustering.py]: continue line w/ normal indent

File: htmresearch/algorithms/hierarchical_clustering.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  cn = ClassifierNetwork()
+  cn.knn = knn
+
+  hc = HierarchicalClustering(cn)
+
+  print hc.getLinkageMatrix().dtype
+  
+  fig = hc.getDendrogram()
+  
+  plt.savefig('test_fig.png')
```
Comment (BoltzmannBrain) [inline htmresearch/algorithms/hierarchical_clustering.py]: add blank line to end of file

File: htmresearch/algorithms/hierarchical_clustering.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
+import numpy
```
Comment (BoltzmannBrain) [inline htmresearch/algorithms/hierarchical_clustering.py]: alpha order for imports

File: htmresearch/algorithms/hierarchical_clustering.py (hunk 1 above)
Comment (BoltzmannBrain) [inline htmresearch/algorithms/hierarchical_clustering.py]: please use double quotes for strings

**Comment by BoltzmannBrain** [comment]
@SaganBolliger nothing major, I left a few line notes.

**Comment by SaganBolliger** [comment]
Implemented suggested changes. Still need to come up with and implement a good visualization for clustering.

**Comment by subutai** [comment]
> Still need to come up with and implement a good visualization for clustering.

That can be a separate PR. No need to wait for that to merge.

**Comment by subutai** [comment]
:+1:

**Pull request merged**